File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    answer = 0
    length = len(triangle)
    
    dp = [[0 for _ in range(length)] for _ in range(length)]
    dp[0][0] = triangle[0][0]
    
    for i in range(1, length):
        for j in range(i + 1):
            if j == 0:
                dp[i][j] = dp[i - 1][0] + triangle[i][j]
            elif j == i:
                dp[i][j] = dp[i - 1][j - 1] + triangle[i][j]
            else:
                dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + triangle[i][j]
    
    return max(dp[-1])

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_set>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    vector<vector<int>> dp(triangle.size(), vector<int>(triangle.size(), 0));
    
    dp[0][0] = triangle[0][0];
    
    for(int i = 1; i < triangle.size(); i++){
        for(int j = 0; j < triangle[i].size(); j++){
            if(j == 0){
                dp[i][j] = dp[i - 1][0] + triangle[i][0];
            }
            else if(j == triangle[i].size() - 1){
                dp[i][j] = dp[i - 1][j - 1] + triangle[i][j];
            }
            else{
                int tmp = triangle[i][j];
                if(dp[i - 1][j - 1] > dp[i - 1][j]){
                    tmp += dp[i - 1][j - 1];
                }
                else{
                    tmp += dp[i - 1][j];
                }
                dp[i][j] = tmp;
            }
        }
    }
    
    for(int i = 0; i < triangle.size(); i++){
        if(answer < dp[dp.size() - 1][i])
            answer = dp[dp.size() - 1][i];
    }
    
    return answer;
}